feat: #7889 merged data-row-edit in de2

jvanveen opened the pull request:
#### Checklist
- [x] Functionality works & meets specifications
- [x] Code reviewed
- [x] Code unit/integration/system tested
- [x] User documentation updated
- [x] Conventional commits (squash if needed)
- [x] No warnings during install
- [x] Updated javascript typing

jvanveen commented:
![](https://github.trello.services/images/mini-trello-icon.png) [#7889 (TBD: As developers we want to combine DE and Data Row Editor so that we can simplify the build/dependencies)](https://trello.com/c/rIowzUKW/410-7889-tbd-as-developers-we-want-to-combine-de-and-data-row-editor-so-that-we-can-simplify-the-build-dependencies)

molgenis-jenkins commented:
:star: PR Preview available on https://preview-frontend-pr-547.dev.molgenis.org

molgenis-jenkins commented:
:star: PR Preview available on https://preview-frontend-pr-547.dev.molgenis.org

MaxPostema commented:
I don't think the "add" in DE2 works correctly there is still a page reload. Editing a row feels nice and fast now! With routing like this we can still use it outside of DE2 (from a app or something) it does feel a bit slower when using it from outside of DE2, but that could be our finicky master server.

I think this works wel. I wonder what the other team members think.

molgenis-jenkins commented:
:star: PR Preview available on https://preview-frontend-pr-547.dev.molgenis.org

molgenis-jenkins commented:
:star: PR Preview available on https://preview-frontend-pr-547.dev.molgenis.org

jvanveen commented:
- Fixed the add button 
- Changed the navigator default link to point to the new router link
- Applied route guards to make sure that all query parameters are filled when entering the MainView
- Applied loading state, so the old data doesn't show up twice after loading another entity because of changing meta-data
- Fixed tests
- Fixed test warnings (use ts 3.9.3 with latest ts-jest)
- Fixes eslint warnings

molgenis-jenkins commented:
:star: PR Preview available on https://preview-frontend-pr-547.dev.molgenis.org

molgenis-jenkins commented:
:star: PR Preview available on https://preview-frontend-pr-547.dev.molgenis.org

MaxPostema commented on packages/data-explorer/src/main.ts:
```diff
@@ -74,18 +74,32 @@ Vue.config.productionTip = false
-  lng: 'en',
-  fallbackLng: 'en',
-  namespace: ['dataexplorer'],
+  lng,
+  fallbackLng,
```
What happens if window.__INITIAL_STATE__ is empty. Maybe we need a in application fallback 'en'?

jvanveen replied:
Good one! Restored the original situation where lng and fallbackLng were set in main.ts

MaxPostema commented on packages/data-explorer/src/main.ts:
```diff
@@ -74,18 +74,32 @@ Vue.config.productionTip = false
 Vue.prototype.$eventBus = new Vue()
 
+// @ts-ignore
+const { lng, fallbackLng } = window.__INITIAL_STATE__
```
Maybe add this line before line #98 "Vue.use" to group it together?

jvanveen replied:
I removed them together with the __INITIAL__STATE in index.html

MaxPostema commented on packages/data-explorer/public/index.html:
```diff
@@ -7,6 +7,13 @@
     <link rel="icon" href="<%= BASE_URL %>favicon.ico">
     <title>data-explorer</title>
     <script>
+    window.__INITIAL_STATE__ = {
```
Having a static __INITIAL_STATE__ feels broken. Can we get thees values from context api?

MaxPostema commented on packages/data-explorer/src/components/DataRowEdit.vue:
```diff
@@ -0,0 +1,267 @@
+  },
+  data () {
+    return {
+      dataExplorerBaseUrl: window.__INITIAL_STATE__.dataExplorerBaseUrl,
```
is this in the context api?

jvanveen replied:
The 'dataExplorerBaseUrl' is actually unused in DataRowEdit. I removed the whole data property.

MaxPostema commented on packages/data-explorer/src/routes.ts:
```diff
@@ -1,26 +1,36 @@
+    name: 'de-view'
+  },
+  {
+    path: '/explorer/:entity/create',
```
Did you not say that you wont change the urls in this commit?

jvanveen replied:
Yes, that was to keep the link in the navigator working. I fixed the link in the navigator now instead.

MaxPostema commented on packages/data-explorer/src/store/actions.ts:
```diff
@@ -26,15 +26,14 @@ export default {
     }
   },
   fetchCardViewData: async ({ commit, state, getters }: { commit: any, state: ApplicationState, getters: any }) => {
-    commit('setLoading', true)
```
Why remove the spinner?

jvanveen replied:
Its wrapped around the whole data fetch function, including the fetchTablePermissions & fetchTableMeta

MaxPostema commented on packages/data-explorer/src/views/MainView.vue:
```diff
@@ -139,21 +139,23 @@ export default Vue.extend({
     this.$eventBus.$off('delete-item')
   },
   async beforeRouteUpdate (to, from, next) {
+    this.setLoading(true)
```
Is this beter here that in the actions?

jvanveen replied:
I think so! :) By putting the loading indicator here, its easier to spot where the loading occurs and where it ends.

MaxPostema commented:
<img width="504" alt="Schermafbeelding 2021-03-05 om 13 44 51" src="https://user-images.githubusercontent.com/46395349/110120143-f0ab1680-7dbc-11eb-8258-475bc0d6faac.png">

I think the title is to close to the menu bar. (chrome)

molgenis-jenkins commented:
:star: PR Preview available on https://preview-frontend-pr-547.dev.molgenis.org

connoratrug commented on packages/data-explorer/src/components/DataRowEdit.vue:
```diff
@@ -0,0 +1,266 @@
+      }
+    },
+    onCancelClick () {
+      this.parent ? this.showParent() : this.goBackToPluginCaller()
+    },
+    goBackToPluginCaller () {
+      window.history.go(-1)
```
should probably vue the router now it it part of the data explorer , ```window.history
``` was more of spike workaround

jvanveen replied:
Refactoring the DataRowEdit component is out of scope for this ticket.

connoratrug replied:
i do not agree , as it now becomes past of the whole DE2 data flow

connoratrug commented on packages/data-explorer/src/components/DataRowEdit.vue:
```diff
@@ -0,0 +1,266 @@
+        }
+      }
+      try {
+        const resp = await repository.fetch(entity, dataRowId)
```
should probably be moved to store to be consistent with data explorer architecture

jvanveen replied:
Refactoring the DataRowEdit component is out of scope for this ticket.

connoratrug commented on packages/data-explorer/src/components/DataRowEdit.vue:
```diff
@@ -0,0 +1,266 @@
+      this.isSaving = false
+      this.alert = { message: alertMsg, type: 'danger' }
+    },
+    errorToMessage (error) {
```
should probably be replaced by more generic data explorer error handler

jvanveen replied:
Refactoring the DataRowEdit component is out of scope for this ticket.

connoratrug commented on packages/data-explorer/src/main.ts:
```diff
@@ -74,18 +74,29 @@ Vue.config.productionTip = false
+router.beforeEach((to, from, next) => {
+  globalThis.defaultRouteQuery = defaultRouteQuery
+  globalThis.to = to.query
+  if (to.name === 'de-view') {
+    const missingQueryParams = Object.keys(defaultRouteQuery).filter((i) => !Object.keys(to.query).includes(i))
+    if (missingQueryParams.length) {
+      next({ path: `/explorer/${to.params.entity}`, query: defaultRouteQuery })
+    } else { next() }
```
why is this code needed , it looks like it trying to hide a issue that should be fixed at the source

jvanveen replied:
It verifies whether all query params are available when it enters the de-view route. Otherwise it redirects with all the required query params.

connoratrug replied:
so then it hides the error making it harder to fix, if a single param is missing , the whole thing now gets replaced with the default , this seems incorrect

connoratrug commented on packages/data-explorer/src/routes.ts:
```diff
@@ -1,26 +1,36 @@
-    path: '/:entity?',
+    path: '/explorer/:entity/:dataRowId/edit',
+    props: true,
+    name: 'de-edit',
```
why the 'de' prefix , suggest to replace with 'edit'

jvanveen replied:
Its useful to prefix actions with a module, so that you don't get name collisions when merging other modules (e.g. the navigator or data-row-edit)

connoratrug replied:
its not a module but a view ( or should be a view)

molgenis-jenkins commented:
:star: PR Preview available on https://preview-frontend-pr-547.dev.molgenis.org

connoratrug commented on packages/data-explorer/src/routes.ts:
```diff
@@ -1,26 +1,36 @@
+    path: '/explorer/:entity',
     component: MainView,
-    name: 'main-view'
+    name: 'de-view'
```
why the 'de'; prefix , the whole app is in de data-explorer namespace , main-view matches the view name and seems more descriptive

jvanveen replied:
See comment above.

connoratrug commented on packages/data-explorer/src/types/ApiResponse.ts:
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable no-use-before-define */
```
this seems like a issue that should no be suppressed

jvanveen replied:
The downgrade to typescript 3.9.3 seemed to fix the eslint parser, but I didn't feel like fixing all the violations in this file. They all rely on each other and its easy to overlook a definition.

connoratrug commented on packages/data-explorer/src/types/RouteQuery.ts:
```diff
@@ -5,4 +5,5 @@ export type RouteQuery = {
     filter?: string // hashed filter object
     hideSidebar?: 'false' | 'true'
     hide?: string // comma separated list of column names
+    view?: string
```
could replace with 'card' | 'table'

jvanveen replied:
Good suggestion, updated!

molgenis-jenkins commented:
:star: PR Preview available on https://preview-frontend-pr-547.dev.molgenis.org

molgenis-jenkins commented:
:star: PR Preview available on https://preview-frontend-pr-547.dev.molgenis.org

connoratrug reviewed (commented):
The PR in its current state merges the data-row-edit code into the data-explorer package, by not integrating it into the data-explorer app architecture a lot of technical dept is added. Would suggest to move 'data-fetch', 'routing' add 'error handling' part of data-row-edit into the data-explorer to avoid different code/architecture styles in a single app. Once merged these are harder to find and update.

molgenis-jenkins commented:
:star: PR Preview available on https://preview-frontend-pr-547.dev.molgenis.org

molgenis-jenkins commented:
:star: PR Preview available on https://preview-frontend-pr-547.dev.molgenis.org

molgenis-jenkins commented:
:star: PR Preview available on https://preview-frontend-pr-547.dev.molgenis.org

molgenis-jenkins commented:
:star: PR Preview available on https://preview-frontend-pr-547.dev.molgenis.org
